Shared/SongReader/Player.swift

```diff
@@ -25,48 +23,62 @@ final class Player: ObservableObject {
     var cancellable = [AnyCancellable]()
 
     func add(url: URL) {
-        guard let context = (UIApplication.shared.delegate as? AppDelegate)?
-            .persistentContainer
-            .viewContext else { return }
-        
-        let newSong = Songs(context: context)
-        newSong.bookmark = try? url.bookmarkData()
-        // TODO: Fix this stuff
-        
-        fetchSong(bookmark: newSong.bookmark)
+        guard url.startAccessingSecurityScopedResource() else {
+            print("Failed to open the file")
+            return
+        }
         
-        try? context.save()
+        container.performBackgroundTask { [self] context in
+            defer { url.stopAccessingSecurityScopedResource() }
+            do {
+                let newSong = Songs(context: context)
+                let bookmark = try url.bookmarkData()
+                newSong.bookmark = bookmark
+                // TODO: Fix this stuff
+                
+                fetchSong(url: url, bookmark: bookmark)
+                
+                
+            } catch {
+                print("An error occured: \(error)")
+            }
+            
+            try? context.save()
```

Comment:
Reason why this is outside of the do/catch?

Reply:
Don't code while tired